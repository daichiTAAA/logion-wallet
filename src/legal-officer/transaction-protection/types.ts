import { UUID } from "../../logion-chain/UUID"

export type LocItemStatus = 'DRAFT' | 'PUBLISHED'

export type LocItemType = 'Data' | 'Document' | 'Linked LOC'

export interface LocItem {
    name: string,
    value: string,
    timestamp: string | null,
    type: LocItemType,
    submitter: string,
    status: LocItemStatus,
    nature?: string,
    target?: UUID,
    newItem: boolean,
}

export enum PublishStatus {
    NONE,
    START,
    PUBLISH_PENDING,
    PUBLISHING,
    PUBLISHED
}

export interface PublishState {
    status: PublishStatus;
}

export interface PublishProps {
    locItem: LocItem
}
