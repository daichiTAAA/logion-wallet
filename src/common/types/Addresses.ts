import moment, { Moment } from 'moment';
import { InjectedAccountWithMeta } from '@polkadot/extension-inject/types';
import { isLegalOfficer } from './LegalOfficer';

export interface Token {
    readonly value: string;
    readonly expirationDateTime: Moment;
}

export interface AccountAddress {
    readonly name: string,
    readonly address: string,
    readonly isLegalOfficer: boolean,
    readonly token?: Token,
}

export default interface Addresses {
    readonly addresses: AccountAddress[],
    readonly currentAddress?: AccountAddress,
}

export type AccountTokens = Record<string, Token>;

export function buildAddresses(
    injectedAccounts: InjectedAccountWithMeta[],
    userAddress: string | undefined,
    tokens: AccountTokens
): Addresses {
    const selectedAddress = currentOrDefaultAddress(injectedAccounts, userAddress, tokens);

    let currentAddress: AccountAddress | undefined;
    const matchingAddress = injectedAccounts.find(injectedAccount => injectedAccount.address === selectedAddress);
    if(selectedAddress !== undefined && matchingAddress !== undefined) {
        currentAddress = {
            name: matchingAddress.meta.name!,
            address: selectedAddress!,
            isLegalOfficer: isLegalOfficer(selectedAddress),
            token: tokenOrUndefinedIfExpired(tokens[selectedAddress]),
        }
    } else {
        currentAddress = undefined;
    }

    return {
        currentAddress,
        addresses: injectedAccounts.map(injectedAccount => ({
            name: injectedAccount.meta.name!,
            address: injectedAccount.address,
            isLegalOfficer: isLegalOfficer(injectedAccount.address),
            token: tokenOrUndefinedIfExpired(tokens[injectedAccount.address]),
        }))
    }
}

function currentOrDefaultAddress(
    injectedAccounts: InjectedAccountWithMeta[],
    currentAddress: string | undefined,
    tokens: AccountTokens
): string | undefined {
    const loggedAddresses = Object.keys(tokens);
    if(currentAddress !== undefined) {
        return currentAddress;
    } else if(loggedAddresses.length > 0) {
        const defaultAccount = injectedAccounts.find(account => loggedAddresses.includes(account.address));
        return defaultAccount?.address;
    } else {
        return undefined;
    }
}

function tokenOrUndefinedIfExpired(token: Token | undefined): Token | undefined {
    if(token === undefined) {
        return undefined;
    }

    const expirationDateTime = token.expirationDateTime;
    if(expirationDateTime.isBefore(moment())) {
        return undefined;
    } else {
        return token;
    }
}
