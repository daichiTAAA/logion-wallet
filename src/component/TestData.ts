import Identity from './types/Identity';
import PostalAddress from './types/PostalAddress';
import { ColorTheme, rgbaToHex } from '../component/ColorTheme';

export const DEFAULT_IDENTITY: Identity = {
    firstName: "John",
    lastName: "Doe",
    email: "john.doe@logion.network",
    phoneNumber: "+1234",
};

export const DEFAULT_ADDRESS: PostalAddress = {
    line1: "Place de le République Française, 10",
    line2: "boite 15",
    postalCode: "4000",
    city: "Liège",
    country: "Belgium",
};

export const COLOR_THEME: ColorTheme = {
    type: 'light',
    shadowColor: rgbaToHex('#3b6cf4', 0.1),
    dashboard: {
        background: '#eff3fe',
        foreground: '#000000',
    },
    sidebar: {
        background: '#ffffff',
        foreground: '#000000',
        activeItemBackground: '#d8e2fd',
    },
    accounts: {
        iconBackground: '#3b6cf4',
        hintColor: rgbaToHex('#000000', 0.6),
        foreground: '#000000',
        background: 'eff3fe',
        legalOfficerIcon: {
            category: 'legal-officer',
            id: 'account-shield'
        }
    },
    frame: {
        background: '#ffffff',
        foreground: '#000000',
        link: rgbaToHex('#3b6cf4', 0.15),
        altBackground: '#eff3fe',
    },
    topMenuItems: {
        iconGradient: {
            from: '#3b6cf4',
            to: '#6050dc',
        }
    },
    bottomMenuItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    },
    buttons: {
        secondary: {
            background: '#eff3fe',
            foreground: '#000000',
        }
    },
    select: {
        background: '#ffffff',
        foreground: '#000000',
        menuBackgroundColor: '#ffffff',
        selectedOptionBackgroundColor: '#ffffff',
    },
    table: {
        background: '#0c163d',
        foreground: '#ffffff',
        header: {
            background: '#0c163d',
            foreground: '#ffffff',
        },
        row: {
            background: '#152665',
            foreground: '#ffffff',
        }
    },
    tabs: {
        background: '#0c163d',
        foreground: '#ffffff',
        inactiveTab: {
            background: '#d8e2fd',
            foreground: '#ffffff',
        },
        borderColor: '#3b6cf4',
    },
    dialog: {
        background: '#eff3fe',
        foreground: '#000000',
        borderColor: '#e6007a',
    },
};
