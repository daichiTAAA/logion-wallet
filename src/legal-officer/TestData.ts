import moment from 'moment';
import { toIsoString } from '../logion-chain/datetime';
import { ProtectionRequest, DEFAULT_LEGAL_OFFICER } from './Types';

export const PENDING_PROTECTION_REQUESTS: ProtectionRequest[] = [
    {
        id: "1",
        requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
        decisions: [
            {
                legalOfficerAddress: DEFAULT_LEGAL_OFFICER,
                status: 'PENDING',
                rejectReason: null,
                decisionOn: null
            }
        ],
        userIdentity: {
            firstName: "John",
            lastName: "Doe",
            email: "john.doe@logion.network",
            phoneNumber: "+1234",
        },
        userPostalAddress: {
            line1: "Place de le République Française, 10",
            line2: "boite 15",
            postalCode: "4000",
            city: "Liège",
            country: "Belgium",
        },
        createdOn: toIsoString(moment('2021-06-10T11:40:00.000Z'))
    }
];

export const PROTECTION_REQUESTS_HISTORY: ProtectionRequest[] = [
    {
        id: "1",
        requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
        decisions: [
            {
                legalOfficerAddress: DEFAULT_LEGAL_OFFICER,
                status: 'ACCEPTED',
                rejectReason: null,
                decisionOn: toIsoString(moment('2021-06-10T11:40:00.000Z'))
            }
        ],
        userIdentity: {
            firstName: "John",
            lastName: "Doe",
            email: "john.doe@logion.network",
            phoneNumber: "+1234",
        },
        userPostalAddress: {
            line1: "Place de le République Française, 10",
            line2: "boite 15",
            postalCode: "4000",
            city: "Liège",
            country: "Belgium",
        },
        createdOn: toIsoString(moment('2021-06-10T11:40:00.000Z'))
    }
];
