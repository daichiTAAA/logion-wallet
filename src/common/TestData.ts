import { Coin, CoinBalance, SYMBOL } from '../logion-chain/Balances';
import { PrefixedNumber, ATTO } from '../logion-chain/numbers';

import Identity from './types/Identity';
import PostalAddress from './types/PostalAddress';
import { Transaction } from './types/ModelTypes';
import { ColorTheme, rgbaToHex } from './ColorTheme';

export const DEFAULT_LEGAL_OFFICER = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"; // Alice
export const ANOTHER_LEGAL_OFFICER = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"; // Bob
export const A_THIRD_LEGAL_OFFICER = "5FLSigC9HGRKVhB9FiEo4Y3koPsNmBmLJbpXg2mp1hXcS59Y"; // Charlie

export const DEFAULT_IDENTITY: Identity = {
    firstName: "John",
    lastName: "Doe",
    email: "john.doe@logion.network",
    phoneNumber: "+1234",
};

export const DEFAULT_ADDRESS: PostalAddress = {
    line1: "Place de le République Française, 10",
    line2: "boite 15",
    postalCode: "4000",
    city: "Liège",
    country: "Belgium",
};

export const COLOR_THEME: ColorTheme = {
    type: 'light',
    shadowColor: rgbaToHex('#3b6cf4', 0.1),
    dashboard: {
        background: '#eff3fe',
        foreground: '#000000',
    },
    sidebar: {
        background: '#ffffff',
        foreground: '#000000',
        activeItemBackground: '#d8e2fd',
    },
    accounts: {
        iconBackground: '#3b6cf4',
        hintColor: rgbaToHex('#000000', 0.6),
        foreground: '#000000',
        background: 'eff3fe',
        legalOfficerIcon: {
            category: 'legal-officer',
            id: 'account-shield'
        }
    },
    frame: {
        background: '#ffffff',
        foreground: '#000000',
        link: rgbaToHex('#3b6cf4', 0.15),
        altBackground: '#eff3fe',
    },
    topMenuItems: {
        iconGradient: {
            from: '#3b6cf4',
            to: '#6050dc',
        }
    },
    bottomMenuItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    },
    recoveryItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    },
    buttons: {
        secondary: {
            background: '#eff3fe',
            foreground: '#000000',
        }
    },
    select: {
        background: '#ffffff',
        foreground: '#000000',
        menuBackgroundColor: '#ffffff',
        selectedOptionBackgroundColor: '#ffffff',
    },
    table: {
        background: '#0c163d',
        foreground: '#ffffff',
        header: {
            background: '#0c163d',
            foreground: '#ffffff',
        },
        row: {
            background: '#152665',
            foreground: '#ffffff',
        }
    },
    tabs: {
        background: '#0c163d',
        foreground: '#ffffff',
        inactiveTab: {
            background: '#d8e2fd',
            foreground: '#ffffff',
        },
        borderColor: '#3b6cf4',
    },
    dialog: {
        background: '#eff3fe',
        foreground: '#000000',
        borderColor: '#e6007a',
    },
};

export const DEFAULT_COIN: Coin = {
    id: 'log',
    iconId: 'log',
    iconType: 'png',
    name: "Logion",
    symbol: SYMBOL
};

export const DEFAULT_COIN_BALANCE: CoinBalance = {
    coin: DEFAULT_COIN,
    balance: new PrefixedNumber("42", ATTO),
    level: 0.1,
};

export const DEFAULT_TRANSACTION: Transaction = {
    from: "from",
    to: "to",
    pallet: "pallet",
    method: "method",
    transferValue: "420",
    tip: "0",
    fee: "125000",
    reserved: "0",
    total: "125420",
    createdOn: "2021-07-28T12:30:00.000",
    type: "Sent",
}
