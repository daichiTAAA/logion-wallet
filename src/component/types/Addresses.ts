import { InjectedAccountWithMeta } from '@polkadot/extension-inject/types';

export interface AccountAddress {
    name: string,
    address: string,
}

export default interface Addresses {
    addresses: AccountAddress[],
    currentAddress: AccountAddress,
}

export function buildAddresses(injectedAccounts: InjectedAccountWithMeta[], userAddress: string): Addresses {
    return {
        currentAddress: {
            name: injectedAccounts.filter(injectedAccount => injectedAccount.address === userAddress)[0].meta.name!,
            address: userAddress,
        },
        addresses: injectedAccounts.map(injectedAccount => { return {
            name: injectedAccount.meta.name!,
            address: injectedAccount.address,
        };})
    }
}
