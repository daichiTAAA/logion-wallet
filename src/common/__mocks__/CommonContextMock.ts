import { TEST_WALLET_USER } from '../../wallet-user/TestData';
import { COLOR_THEME } from '../TestData';
import Addresses, { AccountAddress } from '../types/Addresses';

export let selectAddress = jest.fn();

export let currentAddress = TEST_WALLET_USER;

export function setCurrentAddress(address: string) {
    currentAddress = address;
}

const DEFAULT_ACCOUNT: AccountAddress = {
    name: "name",
    address: TEST_WALLET_USER,
    isLegalOfficer: false,
}

export let addresses: Addresses = {
    currentAddress: DEFAULT_ACCOUNT,
    addresses: [ DEFAULT_ACCOUNT ]
}

export let balances: any = null;

export let transactions: any = null;

export function useCommonContext() {
    return {
        currentAddress,
        selectAddress,
        addresses,
        balances,
        transactions,
        colorTheme: COLOR_THEME,
    };
}

export function setAddresses(value: Addresses) {
    addresses = value;
}

export function setBalances(value: any) {
    balances = value;
}

export function setTransactions(value: any) {
    transactions = value;
}
