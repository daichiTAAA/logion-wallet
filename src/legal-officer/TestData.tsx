import moment from 'moment';
import { toIsoString } from '@logion/node-api/dist/datetime';
import { ProtectionRequest } from '@logion/client/dist/RecoveryClient';

import { DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER } from "../common/TestData";

export const PENDING_PROTECTION_REQUESTS: ProtectionRequest[] = [
    {
        id: "1",
        requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
        legalOfficerAddress: DEFAULT_LEGAL_OFFICER,
        otherLegalOfficerAddress: ANOTHER_LEGAL_OFFICER,
        decision: {
            rejectReason: null,
            decisionOn: null,
        },
        userIdentity: {
            firstName: "John",
            lastName: "Doe",
            email: "john.doe@logion.network",
            phoneNumber: "+1234",
        },
        userPostalAddress: {
            line1: "Place de le République Française, 10",
            line2: "boite 15",
            postalCode: "4000",
            city: "Liège",
            country: "Belgium",
        },
        createdOn: toIsoString(moment('2021-06-10T11:40:00.000Z')),
        isRecovery: false,
        addressToRecover: null,
        status: "PENDING"
    }
];

export const PROTECTION_REQUESTS_HISTORY: ProtectionRequest[] = [
    {
        id: "1",
        requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
        legalOfficerAddress: DEFAULT_LEGAL_OFFICER,
        otherLegalOfficerAddress: ANOTHER_LEGAL_OFFICER,
        decision: {
                rejectReason: null,
                decisionOn: toIsoString(moment('2021-06-10T11:40:00.000Z')),
                locId: "locId"
        },
        userIdentity: {
            firstName: "John",
            lastName: "Doe",
            email: "john.doe@logion.network",
            phoneNumber: "+1234",
        },
        userPostalAddress: {
            line1: "Place de le République Française, 10",
            line2: "boite 15",
            postalCode: "4000",
            city: "Liège",
            country: "Belgium",
        },
        createdOn: toIsoString(moment('2021-06-10T11:40:00.000Z')),
        isRecovery: false,
        addressToRecover: null,
        status: "ACTIVATED"
    }
];

export const RECOVERY_REQUESTS_HISTORY: ProtectionRequest[] = [
    {
        id: "1",
        requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
        legalOfficerAddress: DEFAULT_LEGAL_OFFICER,
        otherLegalOfficerAddress: ANOTHER_LEGAL_OFFICER,
        decision: {
            rejectReason: null,
            decisionOn: toIsoString(moment('2021-06-10T11:40:00.000Z')),
            locId: "locId"
        },
        userIdentity: {
            firstName: "John",
            lastName: "Doe",
            email: "john.doe@logion.network",
            phoneNumber: "+1234",
        },
        userPostalAddress: {
            line1: "Place de le République Française, 10",
            line2: "boite 15",
            postalCode: "4000",
            city: "Liège",
            country: "Belgium",
        },
        createdOn: toIsoString(moment('2021-06-10T11:40:00.000Z')),
        isRecovery: true,
        addressToRecover: "an-address",
        status: "PENDING"
    }
];
