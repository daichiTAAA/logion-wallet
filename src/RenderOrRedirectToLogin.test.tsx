jest.mock('react-router-dom');
jest.mock('./common/CommonContext');

import { shallowRender, act } from './tests';

import RenderOrRedirectToLogin from './RenderOrRedirectToLogin';
import { setAddresses, setCurrentAddress, DEFAULT_USER_ACCOUNT } from './common/__mocks__/CommonContextMock';
import { AccountAddress } from './common/types/Addresses';
import { TEST_WALLET_USER } from './wallet-user/TestData';

test('Given no logged account, when rendering, then redirecting to login with referrer', () => {
    const notLoggedAccount: AccountAddress = {
        name: "name",
        address: TEST_WALLET_USER,
        isLegalOfficer: false,
    };
    setAddresses({
        addresses: [ notLoggedAccount ],
    });
    const render = jest.fn();
    let tree: any;
    act(() => { tree = shallowRender( <RenderOrRedirectToLogin render={ render } /> )});
    expect(tree).toMatchSnapshot();
});

test('Given logged account, when rendering, then rendering', () => {
    setCurrentAddress(DEFAULT_USER_ACCOUNT);
    const render = jest.fn();
    act(() => { shallowRender( <RenderOrRedirectToLogin render={ render } /> )});
    expect(render).toBeCalled();
});
