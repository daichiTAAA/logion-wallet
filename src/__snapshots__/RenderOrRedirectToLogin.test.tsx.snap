// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given no logged account, when rendering, then redirecting to login with referrer 1`] = `
<Redirect
  to={
    Object {
      "pathname": "/login",
      "state": Object {
        "referrer": "/path",
      },
    }
  }
/>
`;
